contracts/proxies/SafeProxy.sol

```diff
@@ -45,9 +45,11 @@ contract SafeProxy {
                 // memory with a 12 byte offset from where the return data starts. Note that we **intentionally** only
                 // do this for the `masterCopy()` call, since the EVM `DELEGATECALL` opcode ignores the most-significant
                 // 12 bytes from the address, so we do not need to make sure the top bytes are cleared when proxying
-                // calls to the `singleton`. This saves us a tiny amount of gas per proxied call.
-                mstore(0x0c, shl(96, _singleton))
-                return(0, 0x20)
+                // calls to the `singleton`. This saves us a tiny amount of gas per proxied call. Additionally, we write
+                // to the "zero-memory" slot, this guarantees that there are no dirty bits in the first 12 bytes in the
+                // return data, which can lead to problems for contracts compiled with Solidity v0.8+.
```

Comment:
Nit: I don't think mentioning the solidity version is worth it here, I believe in general, solidity never guarantees zeroed memory, so maybe rephrasing it as a general statement is better

Reply:
I read the audit report, and it mentioned this:

> The implementation assumes that the first 12 bytes of memory (scratch space) are zeroed out. However, the Solidity compiler does not guarantee this behaviour, even when no code precedes the inline assembly block

It didn't mention anything about the ABI encoding, just the zeroed scratch memory. I think it'd be good to add a link to the source from docs

Reply:
Also, from the docs for solidity v0.7.6:

> uint<M>: enc(X) is the big-endian encoding of X, padded on the higher-order (left) side with zero-bytes such that the length is 32 bytes.

https://docs.soliditylang.org/en/v0.7.6/abi-spec.html#types

Reply:
> this guarantees that there are no dirty bits in the first 12 bytes **in the return data**

This felt implied. I will reword the comment though - it is obviously not as clear as it should be 😅

Reply:
@mmv08 - I updated the wording, let me know if you think it is better now.

Reply:
now it is very clear, thank you!

Reply:
I think the confusion comes from some missing context from the audit readout with Ackee.

One of the questions was "does it even matter if there are dirty bits for the Solidity ABI decoding?". I.e. does Solidity ABI guaranteed to clear the dirty bits from the "shorter" values such as address? And the answer was the with Solidity ABI Coder v2 (default starting v0.8+), it is a hard error to have dirty bits set in the padding, but it wasn't an issue before. Either way, the SOlidity ABI encoding specification is pretty explicit on **not accepting** dirty bits, so we should make sure not to have them regardless of how v0.7 and v0.8 implemented decoding.

I do like how the comment is now better though (and it is both more accurate and informative).